Updating the Rollover page (#2589)
elastic/docs-content: manage-data/lifecycle/index-lifecycle-management/rollover.md

# Rollover [index-rollover]

In {{es}}, the [rollover action](elasticsearch://reference/elasticsearch/index-lifecycle-actions/ilm-rollover.md) replaces your active write index with a new one whenever your index grows too large, too old, or stores too many documents.
This is particularly useful for time-series data, such as logs or metrics, where index growth is continuous and performance can degrade over time.

Without rollover, a single index can grow until searches slow down, shard relocation takes hours, and nodes run low on disk space.

The rollover feature is an important part of how [index lifecycle](../index-lifecycle-management/index-lifecycle.md) and [data stream lifecycles](../data-stream.md) work to keep your indices fast and manageable. By switching the write target of an index, the rollover action provides the following benefits:

> * **Automation** - works with ILM to remove manual index rotation tasks and allows for granular control over retention cycles

Reviewer: "Automation" isn't the best word, I think "Lifecycle" is better. Additionally, we mention both ILM and DLM above, but only ILM here:

```suggestion
* **Lifecycle** - works with lifecycle management (ILM or DLM) to transition the index through its lifecycle actions, and allows for granular control over retention cycles
```

* **Optimized performance** - keeps shard sizes within recommended limits (10-50 GB)
* **Queries run faster** - improves search performance

> Rollover can be triggered via the [API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-indices-rollover) or ILM.

Reviewer: Again we only mention ILM, though DLM also triggers rollover. We should mention it here.

## How rollover works in ILM

> You define a rollover action in the hot phase of an index lifecycle policy. It will run when any of the configured thresholds are met.

Reviewer: ```suggestion
You define a rollover action in the hot phase of an index lifecycle policy. It will run when any of the configured thresholds are met and the write index contains at least one document.
```

You can configure the following rollover conditions:

> * **Size** - an index will rollover when it reaches a set size, for example 50 GB.

Reviewer: ```suggestion
* **Size** - an index will rollover when its shards reaches a set size, for example 50 GB.
```

* **Age** - an index will rollover when an index reaches a certain age, for example 7 days.
* **Document count** - an index will rollover when a shard contains a certain number of documents, for example 2 million.

::::{tip}
Rolling over to a new index based on size, document count, or age is preferable to time-based rollovers. Rolling over at an arbitrary time often results in many small indices, which can have a negative impact on performance and resource usage.
::::

> After rollover occurs in the hot phase of an ILM policy, indices move to other index lifecycle phases like warm, cold, frozen, and delete. Rollover creates a new write index while the old one continues through the lifecycle phases.

Reviewer: This is not necessarily true. There are other ILM actions that occur in the `hot` phase, but _after_ rollover happens. For example, we support the `force_merge` and `searchable_snapshot` action in the hot phase.

**Special rules:**

> * Empty indices rollover is blocked even if they have an associated `max_age` that would otherwise result in a roll over occurring. A policy can override this behavior if you set `min_docs: 0` in the rollover conditions. This can also be disabled on a cluster-wide basis if you set `indices.lifecycle.rollover.only_if_has_documents` to `false`.

Reviewer: ```suggestion
* Rollover for an empty write index is skipped even if they have an associated `max_age` that would otherwise result in a roll over occurring. A policy can override this behavior if you set `min_docs: 0` in the rollover conditions. This can also be disabled on a cluster-wide basis if you set `indices.lifecycle.rollover.only_if_has_documents` to `false`.
```

* Forced rollover occurs if any shard reaches 200 million documents. Usually, a shard will reach 50 GB long before it reaches 200 million documents, but this isn’t the case for space efficient data sets.

## Recommended approaches

Decide your approach to index rotation based on your use case and requirements.

| Use case               | Recommended approach                                      | Benefits                              |
| ---------------------- | --------------------------------------------------------- | ------------------------------------- |
| Logs, metrics          | [Data streams](rollover.md#rollover-data-stream)          | Automatic rollover, minimal setup     |

> | Legacy indexing setup  | [Alias-based rollover](rollover.md#rollover-with-aliases) | Granular control over rollover timing |

Reviewer: The benefits here are a little unclear: data streams do not automatically have rollover unless they are running in Serverless or using DLM. Likewise, both data streams and aliases with backing indices give control over the rollover timing (for on-prem, in Serverless we do not expose this control).

The biggest difference is in the setup, and how the indices are presented to the user.

| Small, static datasets | No rollover                                               | Simpler management                    |

:::{tip}

> For new projects, use data streams. They're simple to manage and handle rollover automatically.

Reviewer: As I mentioned before, data streams do not rollover automatically when ILM is being used, they still require a policy.

:::

### Rotating your indices with data streams [rollover-data-stream]

We recommend using [data streams](../../data-store/data-streams.md) to manage time series data. Data streams automatically track the write index while keeping configuration to a minimum.

> When targeting a data stream, the new backing index becomes the data stream's writing index. The generation of new backing indices is incremented automatically when it reaches a specified age or size.

Reviewer: Another instance of saying it rolls over automatically here (it only does when DLM is being used)

# Rollover [index-rollover]

> In {{es}}, the [rollover action](elasticsearch://reference/elasticsearch/index-lifecycle-actions/ilm-rollover.md) replaces your active write index with a new one whenever your index grows too large, too old, or stores too many documents.

Reviewer 2: ```suggestion
In {{es}}, the [rollover action](elasticsearch://reference/elasticsearch/index-lifecycle-actions/ilm-rollover.md) replaces your active write index with a new one whenever your index grows beyond a specified size, age, or number of documents.
```
Just a thought. I think someone might interpret "too large", "too old", etc. as meaning that Elasticsearch struggles to manage the index in these cases. Perhaps we can make the phrasing more positive, suggesting that ILM gives users control over these dimensions in order to optimize cluster performance and data retention to their needs.

This is particularly useful for time-series data, such as logs or metrics where index growth is continuous, in order to meet performance and retention requirements.

Without rollover, a single index would continue to grow, causing search performance to drop and having a higher administrative burden on the cluster.

> The rollover feature is an important part of how [index lifecycle](../index-lifecycle-management/index-lifecycle.md) (ILM) and [data stream lifecycles](../data-stream.md) (DLM) work to keep your indices fast and manageable. By switching the write target of an index, the rollover action provides the following benefits:

Reviewer 2: @dakrone I think this and line 19 are based off of your [comment](https://github.com/elastic/docs-content/pull/2589#discussion_r2279602382). Do you think it's okay for us to introduce "DLM" here as an acronym for "data stream lifecycle"? I'm a little hesitant since I gather DLM is already in popular use specifically as "data lifecycle management" (e.g. [IBM](https://www.ibm.com/think/topics/data-lifecycle-management), [HP](https://www.hpe.com/ca/en/what-is/data-lifecycle-management.html)), and our use here specifically to denote "data stream lifecycle" doesn't quite match. That is, to my mind "ILM" and "data stream lifecycle" are two different approaches to DLM.

What would you think of our sticking with:
 - index lifecycle management (acronym: "ILM")
 - data stream lifecycle (no official acronym)

If we do introduce an acronym for "data stream lifecycle" we should also [this main page](https://www.elastic.co/docs/manage-data/lifecycle/data-stream) about it.

Reviewer 2: Roger that. Thanks for the clarification! DLM it is. :-)

> The rollover feature is an important part of how [index lifecycle](../index-lifecycle-management/index-lifecycle.md) (ILM) and [data stream lifecycles](../data-stream.md) (DLM) work to keep your indices fast and manageable. By switching the write target of an index, the rollover action provides the following benefits:

Reviewer: ```suggestion
The rollover feature is an important part of how [index lifecycle](../index-lifecycle-management/index-lifecycle.md) ({{ilm-init}}) and [data stream lifecycles](../data-stream.md) ({{dlm-init}}) work to keep your indices fast and manageable. By switching the write target of an index, the rollover action provides the following benefits:
```

> * **Lifecycle** - works with lifecycle management (ILM or DLM) to transition the index through its lifecycle actions and allows for granular control over retention cycles

Reviewer: ```suggestion
* **Lifecycle** - works with lifecycle management ({{ilm-init}} or {{dlm-init}}) to transition the index through its lifecycle actions and allows for granular control over retention cycles
```

* **Optimized performance** - keeps shard sizes within recommended limits (10-50 GB)
* **Queries run faster** - improves search performance

> Rollover can be triggered via the [API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-indices-rollover), ILM, or DLM.

Reviewer: ```suggestion
Rollover can be triggered via the [API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-indices-rollover), {{ilm-init}}, or {{dlm-init}}.
```

> ## How rollover works in ILM

Reviewer: ```suggestion
## How rollover works in {{ilm-init}}
```

> You define a rollover action in the hot phase of an index lifecycle policy. It will run when any of the configured thresholds are met and the write index contains at least one document.

Reviewer: We call these "conditions" in the rollover API

```suggestion
You define a rollover action in the hot phase of an index lifecycle policy. It will run when any of the configured conditions are met and the write index contains at least one document.
```

You can configure the following rollover conditions:

* **Size** - an index will rollover when its shards reach a set size, for example 50 GB.

> * **Age** - an index will rollover when an index reaches a certain age, for example 7 days.

Reviewer 2: ```suggestion
* **Age** - an index will rollover when it reaches a certain age, for example 7 days.
```

* **Document count** - an index will rollover when a shard contains a certain number of documents, for example 2 million.

::::{tip}
Rolling over to a new index based on size, document count, or age is preferable to time-based rollovers. Rolling over at an arbitrary time often results in many small indices, which can have a negative impact on performance and resource usage.
::::

> After rollover, indices move to other index lifecycle phases like warm, cold, frozen, and delete. Rollover creates a new write index while the old one continues through the lifecycle phases.

Reviewer 2: ```suggestion
After rollover, indices move through other configured index lifecycle phases: warm, cold, frozen, and/or delete. Rollover creates a new write index while the old one continues through the lifecycle phases.
```
Would this be better? I'd probably avoid `like` since that seems to me to imply that "warm", "cold", etc. are examples, rather than an exhaustive list.

Author: Cool, that works! 
Why is there an `or` at the end of the list? Are the phases mutually exclusive?

Author: Right, thank you! 🙏

> After rollover, indices move through other configured index lifecycle phases: warm, cold, frozen, and/or delete. Rollover creates a new write index while the old one continues through the lifecycle phases.

Reviewer: I think the `and/or` implies that only delete is the optional phase, but they're all optional, so perhaps just use "and" here:

```suggestion
After rollover, indices move through other configured index lifecycle phases: warm, cold, frozen, and delete. Rollover creates a new write index while the previous one continues through the lifecycle phases.
```

**Special rules:**

> * Rollover for an empty write index is skipped even if they have an associated `max_age` that would otherwise result in a roll over occurring. A policy can override this behavior if you set `min_docs: 0` in the rollover conditions. This can also be disabled on a cluster-wide basis if you set `indices.lifecycle.rollover.only_if_has_documents` to `false`.

Reviewer 2: ```suggestion
* Rollover for an empty write index is skipped even if it has an associated `max_age` that would otherwise result in a rollover occurring. A policy can override this behavior if you set `min_docs: 0` in the rollover conditions. This can also be disabled on a cluster-wide basis if you set `indices.lifecycle.rollover.only_if_has_documents` to `false`.
```

Author: Can't believe I didn't spot this! /0\

Thanks! 😅

## Recommended approaches

Decide your approach to index rotation based on your use case and requirements.

| Use case               | Recommended approach                                      | Setup benefits and limitations                                                                  |
| ---------------------- | --------------------------------------------------------- | ----------------------------------------------------------------------------------------------- |
| Logs, metrics          | [Data streams](rollover.md#rollover-data-stream)          | Configure rollover with lifecycle management, *minimal setup*, control over rollover timing ^1^ |
| Legacy indexing setup  | [Alias-based rollover](rollover.md#rollover-with-aliases) | Configure rollover with lifecycle management, *advanced setup*, control over rollover timing    |
| Small, static datasets | No rollover                                               | Simpler management                                                                              |

> ^1^ Rollover is handled automatically in {{es-serverless}} projects. {applies_to}`serverless: ga`

Reviewer: ```suggestion
^1^ Rollover is handled automatically for data streams in {{es-serverless}} projects. {applies_to}`serverless: ga`
```

> ^1^ Rollover is handled automatically in Serverless projects, therefore configuring rollover timing is abstracted from the user. {applies_to}`serverless: ga`

Author: Roger that! I'm removing the user reference.

:::{tip}

> For new projects, use data streams. They're simple to manage with lifecycle policies where you define phases and actions that handle rollover automatically.

Reviewer: The second sentence reads awkwardly to me:

"They're simple to manage with lifecycle policies where you define phases and actions that handle rollover automatically."

sounds a little unclear about what is defined, and what happens automatically. Perhaps something like:

"Unlike aliases, they're simpler to manage by defining lifecycle actions without requiring additional configuration for rollover."

?

:::

### Rotating your indices with data streams [rollover-data-stream]

> We recommend using [data streams](../../data-store/data-streams.md) to manage time series data. When set up to use an ILM policy that includes rollover, a data stream automatically manages the rotation of your indices. This ensures you can write to the data stream without additional configuration.

Reviewer: I think this could be made more concise:

```suggestion
We recommend using [data streams](../../data-store/data-streams.md) to manage time series data. When set up to use an {{ilm-init}} policy that includes rollover, a data stream manages the rotation of your indices without additional configuration.
```

> We recommend using [data streams](../../data-store/data-streams.md) to manage time series data. When set up to use ILM policies that include rollover, data streams automatically manage the rotation of your indices. This ensures you can write to the data stream without additional configuration.

Reviewer 2: ```suggestion
We recommend using [data streams](../../data-store/data-streams.md) to manage time series data. When set up to use an ILM policy that includes rollover, a data stream automatically manages the rotation of your indices. This ensures you can write to the data stream without additional configuration.
```
Maybe use singular for simplicity and to match the next sentence at line 65.

> When targeting a data stream, the new backing index becomes the data stream's writing index. The generation of new backing indices is incremented automatically when it reaches a specified age or size.

Reviewer 2: ```suggestion
When you use a data stream, each time the current write index reaches a specified age or size, a new backing index is generated, with an incremented number and timestamp, and it becomes the data stream's writing index. 
```
Sorry to belabour this one. Would this phrasing maybe be clearer?

Author: Fixed in a separate commit :)

Each data stream requires an [index template](../../data-store/templates.md) that contains the following:

> * A name or wildcard (`*`) pattern for the data stream.
> * Optional: The mappings and settings applied to each backing index when it’s created.

Reviewer: I think we're missing the `"data_stream": {}` part of the template:

```suggestion
* A name or wildcard (`*`) pattern for the data stream.
* A configuration that indicates a data stream should be used for the index pattern instead of a concrete index.
* Optional: The mappings and settings applied to each backing index when it’s created.
```

For more information about this approach, refer to the [Manage time series data with data streams](../index-lifecycle-management/tutorial-automate-rollover.md#manage-time-series-data-with-data-streams) tutorial.

:::{tip}
Data streams are designed for append-only data, where the data stream name can be used as the operations (read, write, rollover, shrink etc.) target. If your use case requires data to be updated in place, you can perform [update or delete operations directly on the backing indices](../../data-store/data-streams/use-data-stream.md#update-delete-docs-in-a-backing-index).
:::

**Data streams naming pattern**<br>
{{es}} uses a structured naming convention for the backing indices of data streams, following this pattern:

```console
.ds-<DATA-STREAM-NAME>-<yyyy.MM.dd>-<GENERATION>
```

> For more information about data stream naming patterns, refer to the [Generation](../../data-store/data-streams.md#data-streams-generation) section of the Data streams page.

Reviewer 2: ```suggestion
For more information about the data stream naming pattern, refer to the [Generation](../../data-store/data-streams.md#data-streams-generation) section of the Data streams page.
```

### Rotating your indices with aliases [rollover-with-aliases]

 Rotating indices with aliases requires additional configuration steps, including bootstrapping the initial index. For more details about this approach, refer to the [Manage time series data without data streams](../index-lifecycle-management/tutorial-automate-rollover.md#manage-time-series-data-without-data-streams) tutorial.

:::{important}
The use of aliases for rollover requires meeting certain conditions. Review these considerations before applying this approach:

> * The index name must match the pattern `^.-\d+$*`, for example `my-index-000001`.

Author: I based this on the information included [on this page](https://www.elastic.co/docs/reference/elasticsearch/index-lifecycle-actions/ilm-rollover) -- if this formula is wrong, we'd need to update it in the Reference section as well.

I do agree it is a bit too convoluted. I'll simplify it as per your suggestion. Thank you.
